overlord/devicestate/devicestate.go

```diff
@@ -556,7 +598,22 @@ func prereqsFromSnapTaskSet(ts *state.TaskSet) ([]string, error) {
 	return nil, fmt.Errorf("internal error: cannot identify task-snap-setup in taskset")
 }
 
-func remodelTasks(ctx context.Context, st *state.State, current, new *asserts.Model, deviceCtx snapstate.DeviceContext, fromChange string) ([]*state.TaskSet, error) {
+// sideInfoAndPathFromID returns the SideInfo/path for a given snap ID. Note
+// that this will work only for asserted snaps, that is the only case we
+// support for remodeling at the moment.
+func sideInfoAndPathFromID(sis []*snap.SideInfo, paths []string, id string) (*snap.SideInfo, string) {
+	for i, si := range sis {
+		if si.SnapID == id {
+			return sis[i], paths[i]
+		}
+	}
+	return nil, ""
```

Comment:
how are we sure that we don't reach here? won't the callers usually crash on these values? shouldn't this return an error if we reach here instead?

Reply:
The checks happen in `snapstate.UpdateWithDeviceContext` and `snapstate.InstallWithDeviceContext` (checks for localSnapsRequired). The install case was already tested by `TestRemodelTasksSwitchLocalKernelAndGadgetFails`, I have now added a test for the update case (`TestRemodelUC20SwitchKernelGadgetBaseSnapsLocalSnapsFails`).

Reply:
mmh, you meant those methods on remodelOptions? that's the bit I missed

Reply:
Yes, that's correct